### PR TITLE
Crud refactor

### PR DESCRIPTION
**Pull request by jjmocko**
CRUD actions and alternate Response object interactions

### CODE REVIEW

File: lib/teamsnap/collection.rb [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    def delete(client, id)
+      TeamSnap.run(client, :delete, href+"/#{id}", {})
+    end
+
+    def template_attributes
+      template = parsed_collection.fetch(:template) {}
+      data = template.fetch(:data) { [] }
+      data
+        .reject{|col| col.fetch(:name) == "type" }
+        .map{|col| col.fetch(:name) }
```
Comment (PositiveControl) [inline lib/teamsnap/collection.rb]: Spacing nit on the two blocks above:
```
.reject { |col| col.fetch(:name) == "type" }
.map { |col| col.fetch(:name) }
```

**Comment by PositiveControl** [comment]
Almost all files are missing newlines at the end.

File: lib/teamsnap/version.rb [hunk 2]
```diff
@@ -1,3 +1,3 @@
 module TeamSnap
-  VERSION = "1.3.3"
+  VERSION = "2.0.0"
```
Comment (glanotte) [inline lib/teamsnap/version.rb]: do we want to call this `2.0.0.beta`?

File: lib/config/inflecto.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  inflect.irregular "broadcast_sms", "broadcast_smses"
+  inflect.irregular "member_preferences", "member_preferences"
+  inflect.irregular "member_preferences", "members_preferences"
+  inflect.irregular "opponent_results", "opponent_results"
+  inflect.irregular "opponent_results", "opponents_results"
+  inflect.irregular "team_preferences", "team_preferences"
+  inflect.irregular "team_preferences", "teams_preferences"
+  inflect.irregular "team_results", "team_results"
+  inflect.irregular "team_results", "teams_results"
+end
```
Comment (dmccraw) [inline lib/config/inflecto.rb]: EOF

File: lib/config/oj.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Oj.default_options = {
+  :mode => :compat,
+  :symbol_keys => true,
+  :class_cache => true
+}
```
Comment (dmccraw) [inline lib/config/oj.rb]: EOF

File: lib/config/inflecto.rb (hunk 3 above)
Comment (dmccraw) [inline lib/config/inflecto.rb]: Missing 
```
  inflect.irregular "division_member_preferences", "division_members_preferences"
  inflect.irregular "division_preferences", "divisions_preferences"
```

File: lib/teamsnap/api.rb [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    def self.untemplate_attributes(request_attributes)
+      attributes = {}
+      request_attributes.fetch(:template).fetch(:data).each do |datum|
+        attributes[datum.fetch(:name).to_sym] = datum.fetch(:value)
+      end
+      return attributes
+    end
+  end
+end
```
Comment (dmccraw) [inline lib/teamsnap/api.rb]: EOF newline

File: lib/teamsnap/auth_middleware.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    def message(env)
+      env.body || ""
+    end
+  end
+end
+
+Faraday::Request.register_middleware(
+  :teamsnap_auth_middleware => -> { TeamSnap::AuthMiddleware }
+)
```
Comment (dmccraw) [inline lib/teamsnap/auth_middleware.rb]: EOF newline

File: lib/teamsnap/client.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        self,
+        method,
+        klass,
+        sent_args,
+        TeamSnap::Api.template_args?(method)
+      )
+    end
+
+  end
+end
```
Comment (dmccraw) [inline lib/teamsnap/client.rb]: EOF newline

File: lib/teamsnap/collection.rb [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      define_singleton_method(:find) do |client, id|
+        search(client, :id => id).first.tap do |object|
+          raise TeamSnap::NotFound.new(
+            "Could not find a #{self} with an id of '#{id}'."
+          ) unless object
+        end
+      end
+    end
+  end
+end
```
Comment (dmccraw) [inline lib/teamsnap/collection.rb]: EOF newline

File: lib/teamsnap/response.rb [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    def success?
+      @status / 100 == 2
+    end
+
+    def errors?
+      @status / 100 != 2
+    end
+
+  end
+end
```
Comment (dmccraw) [inline lib/teamsnap/response.rb]: EOF newline

File: lib/teamsnap/structure.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        rel_module = collection_module(href, resp, collection)
+
+        TeamSnap.const_set(
+          name, Class.new { include rel_module }
+        ) unless TeamSnap.const_defined?(name, false)
+      end
+
+    end
+  end
+end
```
Comment (dmccraw) [inline lib/teamsnap/structure.rb]: EOF newline

**Comment by PositiveControl** [comment]
After EOF newlines and CI greenage... +1 for :sheep: it

**Pull request merged**